src/MeshKernel/CurvilinearGrid/CurvilinearGrid.cpp

```
@@ -603,20 +603,95 @@ CurvilinearGrid::ComputeDirectionalSmoothingFactors(NodeIndices const& gridpoint
 {
     return CurvilinearGrid(m_gridNodes, m_projection);
 }
+
+double CurvilinearGrid::ComputeAverageNodalDistance(NodeIndices const& index, GridLineDirection direction)
+{
+    if (index.m_m > m_gridNodes.size() || index.m_n > m_gridNodes[0].size())
+    {
+        throw std::invalid_argument("CurvilinearGrid::ComputeAverageNodalDistance: invalid index coordinates");
+    }
+
+    if (direction == GridLineDirection::MDirection)
+    {
+        double numEdges = 0.0;
+        double leftDistance = 0.0;
+        double rightDistance = 0.0;
+        if (index.m_m > 0 && m_gridNodes[index.m_m - 1][index.m_n].IsValid())
+        {
+            leftDistance = ComputeDistance(m_gridNodes[index.m_m][index.m_n], m_gridNodes[index.m_m - 1][index.m_n], m_projection);
+            numEdges += 1.0;
+        }
+        if (index.m_m + 1 < m_gridNodes.size() && m_gridNodes[index.m_m + 1][index.m_n].IsValid())
+        {
+            rightDistance = ComputeDistance(m_gridNodes[index.m_m][index.m_n], m_gridNodes[index.m_m + 1][index.m_n], m_projection);
+            numEdges += 1.0;
+        }
+        return (leftDistance + rightDistance) / numEdges;
+    }
+    if (direction == GridLineDirection::NDirection)
+    {
+        double numEdges = 0.0;
+        double bottomDistance = 0.0;
+        double upDistance = 0.0;
+        if (index.m_n > 0 && m_gridNodes[index.m_m][index.m_n - 1].IsValid())
+        {
+            bottomDistance = ComputeDistance(m_gridNodes[index.m_m][index.m_n], m_gridNodes[index.m_m][index.m_n - 1], m_projection);
+            numEdges += 1.0;
+        }
+        if (index.m_n + 1 < m_gridNodes[0].size() && m_gridNodes[index.m_m][index.m_n + 1].IsValid())
+        {
+            upDistance = ComputeDistance(m_gridNodes[index.m_m][index.m_n], m_gridNodes[index.m_m][index.m_n + 1], m_projection);
+            numEdges += 1.0;
+        }
+        return (bottomDistance + upDistance) / numEdges;
+    }
+
+    return doubleMissingValue;
```

it looks like this return value should be unreachable, because "direction" will be one of the 2 cases handled above. you could throw an exception before the return statement to make it more explicit we're not supposed to end up at that line

---

Done